extensions/amp-video/0.1/mixins.js

```diff
@@ -0,0 +1,524 @@
+  }
+
+  /** @override */
+  toggleControls(show) {
+    this.video.controls = show;
+  }
+
+  /** @override */
+  fullscreenEnter() {
+    fullscreenEnter(dev().assertElement(this.video));
+  }
+
+  /** @override */
+  fullscreenExit() {
+    fullscreenExit(dev().assertElement(this.video));
+  }
+
+  /** @override */
+  isFullscreen() {
+    return isFullscreenElement(dev().assertElement(this.video));
+  }
+
+  /** @override */
+  getCurrentTime() {
+    return this.video.currentTime;
+  }
+
+  /** @override */
+  getDuration() {
+    return this.video.duration;
+  }
+
+  /** @override */
+  getPlayedRanges() {
+    return VideoUtils.getPlayedRanges(
+        /** @type {!HTMLMediaElement} */ (dev().assert(this.video)));
+  }
+}
+
+
+// TODO(alanorozco): This mixin should not inherit from `VideoElementMixin`.
+// This is blocked by mediapool control coming from `AmpStoryPage`.
+export class MediaPoolVideoMixin extends VideoElementMixin {
+
+  /** @param {!./amp-video.AmpVideo} impl */
+  constructor(impl) {
+    super(impl);
+
+    const {element} = impl;
+
+    /** @private {?Promise<!Element>} */
+    this.placeholder_ = null;
+
+    /** @private {boolean} */
+    this.isAllocated_ = false;
+
+    /** @private {?../../amp-story/0.1/media-pool.MediaPool} */
+    this.pool_ = null;
+
+    /** @private {?../../amp-story/0.1/media-pool.MediaInfoDef} */
+    this.mediaInfo_ = EMPTY_MEDIA_INFO;
+
+    /** @private @const {!Promise<!../../amp-story/0.1/media-pool.MediaPool>} */
+    this.poolPromise_ = Services.mediaPoolFor(element).then(service =>
+      (this.pool_ = service.poolFor(element)));
+  }
+
+  /** @override */
+  isVideoSupported() {
+    return true; // Assumption.
+  }
+
+  /** @override */
+  installEventHandlers(unusedEventsToForward, setMuted) {
+    // No need to forward events as MediaPool re-dispatches all playback events
+    // from parent element.
+
+    const {element} = this.impl;
+
+    listen(element, MediaPoolEvents.ALLOCATED, () => {
+      this.isAllocated_ = true;
+      this.hidePlaceholder();
+    });
+
+    listen(element, MediaPoolEvents.DEALLOCATED, e => {
+      const {detail} = e;
+      this.isAllocated_ = false;
+      this.mediaInfo_ =
+        /** @type {!../../amp-story/0.1/media-pool.MediaInfoDef} */ (detail);
+      this.showPlaceholder();
+    });
+
+    listen(element, 'volumechange', () => {
+      setMuted(this.getMediaInfo_('muted'));
+    });
+  }
+
+  /** @visibleForTesting */
+  showPlaceholder() {
+    if (this.placeholder_) {
+      this.togglePlaceholder_(/* show */ true);
+      return this.placeholder_;
+    }
+
+    const {element} = this.impl;
+    const posterSrc = dev().assertString(element.getAttribute('poster'));
+
+    const img = new Image();
+    const {impl} = this;
+    const {win} = this.impl;
+
+    const placeholder = createElementWithAttributes(win.document, 'div', dict({
+      'class': 'i-amphtml-placeholder',
+    }));
+
+    img.src = posterSrc;
```

Comment:
Doesn't `loadPromise` (below) listen for events?  If so, shouldn't this be after `loadPromise`?  To avoid the `load` event from firing before the `loadPromise` is listening?